auto-increment.md

```diff
@@ -0,0 +1,211 @@
+```
+
+```sql
+mysql> select * from t;
++----+---+
+| id | c |
++----+---+
+| 1  | 1 |
+| 2  | 2 |
+| 3  | 3 |
+| 4  | 4 |
+| 5  | 5 |
++----+---+
+5 rows in set (0.01 sec)
+```
+
+In addition, `AUTO_INCREMENT` also supports the `INSERT` statements that explicitly specify column values. In such cases, TiDB stores the explicitly specified values:
+
+{{< copyable "sql" >}}
+
+```sql
+insert into t(id, c) values (6, 6);
+```
+
+```sql
+mysql> select * from t;
++----+---+
+| id | c |
++----+---+
+| 1  | 1 |
+| 2  | 2 |
+| 3  | 3 |
+| 4  | 4 |
+| 5  | 5 |
+| 6  | 6 |
++----+---+
+6 rows in set (0.01 sec)
+```
+
+The usage above is the same as that of `AUTO_INCREMENT` in MySQL. However, in terms of the specific value that is implicitly assigned, TiDB differs from MySQL significantly.
+
+## Working principles
+
+TiDB implements the `AUTO_INCREMENT` implicit assignment in the following way:
+
+For each auto-increment column, a globally visible key-value pair is used to record the maximum ID that has been assigned. In a distributed environment, communication between nodes has some overhead. Therefore, to avoid the issue of write amplification, each TiDB node applies for a batch of consecutive IDs as cache when assigning IDs, and then applies for the next batch of IDs after the first batched is assigned. TiDB nodes do not have to apply to the storage node for IDs at each assigning. For example:
+
+```sql
+create table t(id int unique key AUTO_INCREMENT, c int);
+```
+
+Assume two TiDB instances, `A` and `B`, in the cluster. If you execute an `INSERT` statement for the `t` table on `A` and `B` respectively:
+
+```sql
+insert into t (c) values (1)
+```
+
+Instance `A` might cache the auto-increment ID of `[1,30000]`, and instance `B` might cache the auto-increment ID of `[30001,60000]`. In the future `INSERT` statements, these cached ID of each instance will be assigned to the `AUTO_INCREMENT` column as the default value.
+
+## Basic Features
+
+### Uniqueness
+
+> **Warning:**
+>
+> When the cluster has multiple TiDB instances, if the table schema contains the auto-increment ID, it is recommended not to use explicit insert and implicit assignment at the same time (i.e. using the default value of the auto-increment column and the custom value). Otherwise, it might break the uniqueness of implicitly assigned values.
+
+In the example above, perform the following operations in order:
+
+1. The client inserts a statement `insert into t values (2, 1)` to instance `B`, which sets `id` to `2`. The statement is successfully executed.
+
+2. The client sends a statement `insert into t (c) (1)` to instance `A`. This statement does not specify the value of `id`, so the ID is assigned by `A`. At present, because `A` caches the ID of `[1, 30000]`, it might assign `2` as the value of the auto-increment ID, and increases the local counter by `1`. At this time, the data whose ID is `2` already exists in the database, so the `Duplicated Error` error is returned.
```

Comment:
```suggestion
2. The client sends a statement `insert into t (c) (1)` to instance `A`. This statement does not specify the value of `id`, so the ID is assigned by `A`. At present, because `A` caches the IDs of `[1, 30000]`, it might assign `2` as the value of the auto-increment ID, and increases the local counter by `1`. At this time, the data whose ID is `2` already exists in the database, so the `Duplicated Error` error is returned.
```